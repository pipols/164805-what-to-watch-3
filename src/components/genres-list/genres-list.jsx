import React from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {ActionType} from "../../reducer";
import {mapGenresToFilter} from "../../const/genres";

const MAX_GENRES = 9;
const DEFAULT_FILTER = `All genres`;
// уникальные жанры из фильмов
const getUniqueGenres = (films) => {
  const genresList = films.map((film) => film.genre);
  const uniqueGenres = new Set(genresList);
  return [...uniqueGenres].slice(0, MAX_GENRES);
};

const getFilters = (films) => getUniqueGenres(films).map((genre) => mapGenresToFilter.get(genre));

const GenresList = ({films, onFilterClick}) => {
  const filters = getFilters(films);
  filters.unshift(DEFAULT_FILTER);

  return (
    <ul className="catalog__genres-list">
      {filters.map((filter) =>
        <li key={filter} className="catalog__genres-item">
          <a
            onClick={(evt) => {
              evt.preventDefault();
              onFilterClick(filter);
            }}
            href="#"
            className="catalog__genres-link">{filter}
          </a>
        </li>
      )}
    </ul>
  );
};

GenresList.propTypes = {
  films: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    poster: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired
  })).isRequired,
  onFilterClick: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  films: state.films
});

const mapDispatchToProps = (dispatch) => ({
  onFilterClick(filter) {
    dispatch({type: ActionType.SET_GENRE, payload: filter});
  }
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
