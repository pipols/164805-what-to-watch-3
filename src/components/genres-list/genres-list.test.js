import React from "react";
import renderer from "react-test-renderer";
import {GenresList} from "./genres-list.jsx";

const onFilterClick = () => {};
const onItemClick = () => {};
const filters = [`Comedies`, `Crime`];

it(`<GenresList/> renders`, () => {
  const tree = renderer
    .create(<GenresList
      filters={filters}
      onFilterClick={onFilterClick}
      onItemClick={onItemClick}
      activeItem={filters[0]}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
