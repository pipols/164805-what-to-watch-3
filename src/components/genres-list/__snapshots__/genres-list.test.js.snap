// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenresList/> renders 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      All genres
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      Comedies
    </a>
  </li>
</ul>
`;
